docs/concepts-and-explanations/javatemplate.md

```diff
@@ -120,6 +120,51 @@ JavaTemplate.builder("Duration.ofMillis(#{any(int)})")
 There are many examples of `JavaTemplate` stubs in [rewrite-testing-frameworks](https://github.com/openrewrite/rewrite-testing-frameworks) and [rewrite-spring](https://github.com/openrewrite/rewrite-spring).
 :::
 
+## Context-free vs. Context-sensitive JavaTemplates
+
+There are two types of JavaTemplates – each having their own advantages/disadvantages:
+
+A **context-free** JavaTemplate does not depend on the surrounding LST. It is a static template that simply replaces or inserts code without considering the broader context of the source file. This is particularly useful for simple transformations like replacing method calls or inserting boilerplate code.
+
+For instance, the following template, when applied, would not check the existing imports or variables in scope:
+
+```java
+JavaTemplate.builder("System.out.println(\"Hello, World!\");")
+            .build();
+```
+
+The main advantage of context-free templates is that they only need to be parsed once before their parameters can be substituted. This has a much smaller performance impact on recipe runs.
+
+A **context-sensitive** JavaTemplate, on the other hand, is aware of and adapts to the surrounding LST. It can capture elements from the existing code, such as variable names, types, or method parameters.
+
+An example can help this make more sense. Let's presume we have code that looks like this:
+
+```java
+public void logMessage() {
+    String message = "Application started";
+    System.out.println(message);
+}
+```
+
+We want to change the `System.out.println` to be a logging method, but we don't want to use the variable `message`. We could write a template like:
+
+```java
+JavaTemplate.builder("logger.info(#{any(String)});")
```

Comment:
This template does actually not have to be declared as context-sensitive. That is because also context-free templates can have parameters, like the `#{any(String)}` in this example. The context-sensitive template for this example would have been: `"logger.info(message);"`. Maybe that example is however a bit too simplistic?

Reply:
Hmm I get what you're saying - but ya I want to demonstrate the substitution. Do you have any ideas for a better example?

As I'm not as involved in recipes as you are, I'm happy to defer to your expertise here - whether that be putting it as just `"logger.info(message);"` or finding a better example.

Reply:
No immediate better examples come to mind; indeed when referencing a field and it's somehow awkward to get a type attributed reference to insert into the template is when contextSensitive makes sense.